core/src/main/java/com/linecorp/armeria/client/websocket/WebSocketSession.java

```diff
@@ -0,0 +1,119 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.websocket;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.reactivestreams.Publisher;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.stream.PublisherBasedStreamMessage;
+import com.linecorp.armeria.common.stream.StreamMessage;
+import com.linecorp.armeria.common.websocket.WebSocket;
+import com.linecorp.armeria.common.websocket.WebSocketFrame;
+import com.linecorp.armeria.internal.common.websocket.WebSocketFrameEncoder;
+
+/**
+ * A WebSocket session that is created after {@link WebSocketClient#connect(String)} succeeds.
+ * You can start sending {@link WebSocketFrame}s via {@link #setOutbound(WebSocket)}. You can also subscribe to
+ * {@link #inbound()} to receive {@link WebSocketFrame}s from the server.
+ */
+public final class WebSocketSession {
+
+    private final ClientRequestContext ctx;
+    private final ResponseHeaders responseHeaders;
+    @Nullable
+    private final String subprotocol;
+    private final WebSocket inbound;
+    private final CompletableFuture<StreamMessage<HttpData>> outboundFuture;
+    private final WebSocketFrameEncoder encoder;
+
+    WebSocketSession(ClientRequestContext ctx, ResponseHeaders responseHeaders, WebSocket inbound,
+                     CompletableFuture<StreamMessage<HttpData>> outboundFuture,
+                     WebSocketFrameEncoder encoder) {
+        this.ctx = ctx;
+        this.responseHeaders = responseHeaders;
+        subprotocol = responseHeaders.get(HttpHeaderNames.SEC_WEBSOCKET_PROTOCOL);
+        this.inbound = inbound;
+        this.outboundFuture = outboundFuture;
+        this.encoder = encoder;
+    }
+
+    /**
+     * Returns the {@link ClientRequestContext}.
+     */
+    public ClientRequestContext context() {
+        return ctx;
+    }
+
+    /**
+     * Returns the {@link ResponseHeaders}.
+     */
+    public ResponseHeaders responseHeaders() {
+        return responseHeaders;
+    }
+
+    /**
+     * Returns the subprotocol negotiated between the client and the server.
+     */
+    @Nullable
+    public String subprotocol() {
+        return subprotocol;
+    }
+
+    /**
+     * Returns the {@link WebSocket} which is used to receive WebSocket frames from the server.
+     */
+    public WebSocket inbound() {
+        return inbound;
+    }
+
+    /**
+     * Sets the {@link WebSocket} which is used to send WebSocket frames to the server.
+     */
+    public boolean setOutbound(Publisher<? extends WebSocketFrame> outbound) {
```

Comment:
Do we use `set` in our setters instead of the record style?
```suggestion
    public boolean outbound(Publisher<? extends WebSocketFrame> outbound) {
```

Also, do you think it would be confusing to add a shortcut in case users just want to write directly?

```java
val writer = WebsocketSession.outbound() // creates a writer internally and sets to `outboundFuture`
writer.write("hi")
```

Reply:
> Do we use set in our setters instead of the record style?

I believe so and this is one of examples: https://github.com/line/armeria/blob/7da39b8b24f80bb24ac7f7bfd040c995a1ee3ed5/core/src/main/java/com/linecorp/armeria/client/ClientRequestContext.java#L307-L323 😉